src/N3Reasoner.js

```diff
@@ -0,0 +1,206 @@
+    const newRules = [];
+
+    function addRule(conclusion) {
+      if (conclusion.next)
+        conclusion.next.forEach(rule => {
+          newRules.push([conclusion.subject.value, conclusion.predicate.value, conclusion.object.value, rule]);
+        });
+    }
+
+    // eslint-disable-next-line func-style
+    const addConclusions = conclusion => {
+      conclusion.forEach(c => {
+        // eslint-disable-next-line max-nested-callbacks
+        this._add(c.subject.value, c.predicate.value, c.object.value, content, () => { addRule(c); });
+      });
+    };
+
+    this._evaluateRules(rules, content, addRule);
+
+    let r;
+    while ((r = newRules.pop()) !== undefined) {
+      const [subject, predicate, object, rule] = r;
+      const v1 = rule.basePremise.subject.value;
+      if (!v1) rule.basePremise.subject.value = subject;
+      const v2 = rule.basePremise.predicate.value;
+      if (!v2) rule.basePremise.predicate.value = predicate;
+      const v3 = rule.basePremise.object.value;
+      if (!v3) rule.basePremise.object.value = object;
+
+      if (rule.premise.length === 0) {
+        addConclusions(rule.conclusion);
+      }
+      else {
+        this._evaluatePremise(rule, content, addRule);
+      }
+
+      if (!v1) rule.basePremise.subject.value = null;
+      if (!v2) rule.basePremise.predicate.value = null;
+      if (!v3) rule.basePremise.object.value = null;
+    }
+  }
+
+  _createRule({ premise, conclusion }) {
+    const ids = this._store._ids, entities = this._store._entities, varMapping = {};
+
+    const toId = value => value.termType === 'Variable' ?
+      // If the term is a variable then create an empty object that values can be placed into
+      (varMapping[value.value] = varMapping[value.value] || {}) :
+      // If the term is not a variable then set the ID value
+      { value: ids[value = termToId(value)] || (ids[entities[++this._store._id] = value] = this._store._id) };
+
+    // eslint-disable-next-line func-style
+    const t = term => ({ subject: toId(term.subject), predicate: toId(term.predicate), object: toId(term.object) });
+
+    return {
+      premise: premise.map(p => t(p)),
+      conclusion: conclusion.map(p => t(p)),
+      variables: Object.values(varMapping),
+    };
+  }
+
+  reason(rules) {
+    if (!Array.isArray(rules)) {
+      rules = getRulesFromDataset(rules);
+    }
+    rules = rules.map(rule => this._createRule(rule));
+
+    for (const r1 of rules) {
+      for (const r2 of rules) {
+        for (let i = 0; i < r2.premise.length; i++) {
+          const p = r2.premise[i];
+          for (const c of r1.conclusion) {
+            if (termEq(p.subject, c.subject) && termEq(p.predicate, c.predicate) && termEq(p.object, c.object)) {
+              const set = new Set();
+
+              const premise = [];
+
+              // Since these *will* be substited when we apply the rule,
+              // we need to do this so that we index correctly in the subsequent section
```

Comment:
```suggestion
              // Since these *will* be substituted when we apply the rule,
              // we need to do this, so that we index correctly in the subsequent section
```